assert/cmp/compare.go

```diff
@@ -15,6 +15,30 @@ import (
 // with details about why it failed.
 type Comparison func() (success bool, message string)
 
+// ComparisonWithResult is a Comparison that returns a Result object.
+type ComparisonWithResult func() Result
+
+// Result of a Comparison
+type Result struct {
```

Comment:
I wonder if `Result` should be an interface instead. It could be something like this:

```go
type Result interface {
    Success() bool
    Format(opts FormatOpts) string
}

type FormatOpts struct {
    Args []ast.Expr
}
```
It would mean that `renderMessage` could move to `cmp`, and it would be easier to test comparisons directly without needing the `assert` package. It would also make it easier for the `assert` package to provide more data to format in a backwards compatible way (by adding to `FormatOpts`). However, it seems very unlikely that there is any other information to provide.

`renderMessage` could move to `cmp` either way, make it easier to test.